File: php/allergies/Allergies.php
<?php

class Allergen
{
    public const EGGS = 1;
    public const PEANUTS = 2;
    public const SHELLFISH = 4;
    public const STRAWBERRIES = 8;
    public const TOMATOES = 16;
    public const CHOCOLATE = 32;
    public const POLLEN = 64;
    public const CATS = 128;

    public function __construct(private int $score)
    {
    }
    public static function allergenList(): array
    {
        return array_map(static fn ($allergen) => new Allergen($allergen),
            (new ReflectionClass(__CLASS__))->getConstants());
    }
    public function getScore(): int
    {
        return $this->score;
    }
}

class Allergies
{
    public function __construct(private int $mask)
    {
    }

    public function isAllergicTo(Allergen $allergen): bool
    {
        return $this->mask & $allergen->getScore();
    }
    public function getList(): array
    {
        return array_filter(Allergen::allergenList(), fn ($allergen) => $this->isAllergicTo($allergen));
    }
}

File: erlang/allergies/src/allergies.erl
-module(allergies).
-export([allergies/1, is_allergic_to/2]).
-define(DICT, #{ eggs=>1,
					    peanuts=>2,
					    shellfish=>4,
					    strawberries=>8,
					    tomatoes=>16,
					    chocolate=>32,
					    pollen=>64,
					    cats=>128}).


is_allergic_to(Substance, Score) -> Score band maps:get(Substance, ?DICT) > 0.

allergies(Score) -> [
  K || K <- maps:keys(?DICT),
  is_allergic_to(K, Score)
].

